📖 book: add Writing a ClusterClass doc (#5997)
kubernetes-sigs/cluster-api: docs/book/src/tasks/experimental-features/cluster-class/write-clusterclass.md

# Writing a ClusterClass

While a ClusterClass can be also used for a single Cluster, usually a ClusterClass is used across many
clusters. Therefore a ClusterClass becomes more useful when it's flexible enough to support a wide
range of use cases. The goal of this document is to explain how ClusterClasses can be written in

> general and how they can be written in a way that they are flexible enough to support many use cases.

Author: Thx! I'll take your variant.

kind: ClusterClass
metadata:
  name: docker-clusterclass-v0.1.0
spec:
  controlPlane:
    ref:
      apiVersion: controlplane.cluster.x-k8s.io/v1beta1
      kind: KubeadmControlPlaneTemplate
      name: docker-clusterclass-v0.1.0
      namespace: default
    machineInfrastructure:
      ref:
        kind: DockerMachineTemplate
        apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
        name: docker-clusterclass-v0.1.0
        namespace: default
  infrastructure:
    ref:
      apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
      kind: DockerClusterTemplate
      name: docker-clusterclass-v0.1.0-control-plane
      namespace: default
  workers:
    machineDeployments:
    - class: default-worker
      template:
        bootstrap:
          ref:
            apiVersion: bootstrap.cluster.x-k8s.io/v1beta1
            kind: KubeadmConfigTemplate
            name: docker-clusterclass-v0.1.0-default-worker
            namespace: default
        infrastructure:
          ref:
            apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
            kind: DockerMachineTemplate
            name: docker-clusterclass-v0.1.0-default-worker
            namespace: default
```

> The following example shows a Cluster using this ClusterClass. In this case a `KubeadmControlPlane` 

Reviewer: Might be easier for readers to follow along if there are  example templates here so users can follow this guide to create and edit a working ClusterClass and Cluster using the yamls here.

Reviewer 2: we can eventually link to the full yaml, but I see value also in keeping thing simple and short

Author: My idea was to essentially only make it a series of snippets to show the relevant concepts.
Otherwise it would be so much YAML to dig through and we would have to keep maintaining them.

Author: What do we think about referencing the CAPD ClusterClass cluster-template that we are also publishing here?

Reviewer 2: @sbueringer  FYI @ykakarap have found a way to provide collapsible sections with the full yaml, we can probably use the same approach here as well

Author: At the moment I've added a link to the published YAMLs. I'm not sure if we want to have too many copies of CAPD ClusterClasses and Clusters, especially as probably nobody tested them and it's hard to ensure manually the full examples keep working.

with the corresponding `DockerMachineTemplate`, a `DockerCluster` and a `MachineDeployment` with 
the corresponding `KubeadmConfigTemplate` and `DockerMachineTemplate` will be created. This basic 
ClusterClass is already very flexible. Via the topology on the Cluster the following can be configured:
* `.spec.topology.version`: the Kubernetes version of the Cluster

> * `.spec.topology.controlPlane`: control plane replicas and their metadata

Reviewer: nit 
```suggestion
* `.spec.topology.controlPlane`: ControlPlane replicas and their metadata
```

Author: I was using that more as the "word" control plane vs a reference to the type. Like we would write control plane machines

Author: At some point we should probably take a look at the Kubernetes style guide about using API types in documentation. It's very extensive.

It's like Pandora's box, but here is the link: https://kubernetes.io/docs/contribute/style/style-guide/ :joy:

Reviewer: It's a tough one, and inconsequential, but it sticks out when we're using MachineDeployment and control plane in the same sentence.

Author: Hm yeah. I'll use ControlPlane (done)

* `.spec.topology.workers`: MachineDeployments and their replicas, metadata and failure domain
    * **Note**: It's possible to create MachineDeployments with different configurations, by
      defining multiple MachineDeployment classes in the ClusterClass. 

```yaml
apiVersion: cluster.x-k8s.io/v1beta1
kind: Cluster
metadata:
  name: my-docker-cluster
spec:
  topology:
    class: docker-clusterclass
    version: v1.22.4
    controlPlane:
      replicas: 3
      metadata:

>         labels: {}

Reviewer: It might be clearer to add  a sample label / annotation here to show usage - same below.

Author: My idea was to keep it short, but I added labels now. If you've got a good idea for labels/annotations please let me know :)

        annotations: {}
    workers:
      machineDeployments:
      - class: default-worker
        name: md-0
        replicas: 4
        metadata:
          labels: {}
          annotations: {}
        failureDomain: region
```

Best practices:
* The ClusterClass name should be generic enough to make sense across multiple clusters, i.e. a 
  name which corresponds to a single Cluster, e.g. "my-cluster", is not recommended.
* Try to keep the ClusterClass names short and consistent (if you publish multiple ClusterClasses).
* As a ClusterClass usually evolves over time and you might want to rebase Clusters from one version 
  of a ClusterClass to another, consider including a version suffix in the ClusterClass name.
  For more information about changing a ClusterClass please see: [Changing a ClusterClass].
* Prefix the templates used in a ClusterClass with the name of the ClusterClass.
* Don't use templates with the same name in multiple ClusterClasses. This is automatically taken care
  of by prefixing the templates with the name of the ClusterClass.

<aside class="note">

<h1>Tip: clusterctl alpha topology plan</h1>

> `clusterctl alpha topology plan` can be used to render objects of a managed topology without having to create or update
> a Cluster or a ClusterClass in Kubernetes. For more details please see: [clusterctl alpha topology plan].

Reviewer 2: What about
```suggestion
The `clusterctl alpha topology plan` command can be used to test ClusterClasses; the output will show
you how the resulting Cluster will look like, but without actually creating it.
For more details please see: [clusterctl alpha topology plan].
```

Author: Sounds good

As shown above, basic ClusterClasses are already very powerful. But there are cases where 
more powerful mechanisms are required. Let's assume you want to manage multiple Clusters 
with the same ClusterClass, but they require different values for a field in one of the 
referenced templates of a ClusterClass. A concrete example would be to deploy Clusters in 
different AWS regions. In this case, every cluster needs a Cluster-specific value for 
`.spec.region` in `AWSCluster`. Use cases like this can be implemented with ClusterClass 
patches.

**Defining variables in the ClusterClass**

The following example shows how variables can be defined in the ClusterClass.
A variable definition specifies the name and the schema of a variable and if it is 
required. The schema defines how a variable is defaulted and validated. It supports 
a subset of the schema of CRDs. For more information please see the [godoc](https://doc.crds.dev/github.com/kubernetes-sigs/cluster-api/cluster.x-k8s.io/ClusterClass/v1beta1#spec-variables-schema-openAPIV3Schema).

```yaml
apiVersion: cluster.x-k8s.io/v1beta1
kind: ClusterClass
metadata:
  name: docker-clusterclass-v0.1.0
spec:
  ...
  variables:

>   - name: imageRepository

Author: I can, but I tried to use something which works with CAPD when possible.

I can switch over to AWS (and in the other snippets in `## ClusterClass with patches`) if we want to.

>   - name: imageRepository

Reviewer 2: I don't have strong preferences about the use case, but if we can keep both consistent I think it will be greatly beneficial for the users

>   - name: imageRepository

Author: I think I would prefer staying with CAPD in the examples. While AWS makes for a better example to explain real-life use cases, CAPD is something that users can easily test (locally) and play around with, without having to create a whole AWS cluster.

>   - name: imageRepository

Author: I'll rewrite the example text to match it with the imageRepository use case

please see the [godoc](https://doc.crds.dev/github.com/kubernetes-sigs/cluster-api/cluster.x-k8s.io/ClusterClass/v1beta1#spec-patches-definitions).

```yaml
apiVersion: cluster.x-k8s.io/v1beta1
kind: ClusterClass
metadata:
  name: docker-clusterclass-v0.1.0
spec:
  ...
  patches:
  - name: imageRepository
    definitions:
    - selector:
        apiVersion: controlplane.cluster.x-k8s.io/v1beta1
        kind: KubeadmControlPlaneTemplate
        matchResources:
          controlPlane: true
      jsonPatches:
      - op: add
        path: /spec/template/spec/kubeadmConfigSpec/clusterConfiguration/imageRepository
        valueFrom:
          variable: imageRepository
```

**Notes**:
* Only fields below `/spec` can be patched.
* Only `add`, `remove` and `replace` operations are supported.
* It's only possible to append and prepend to arrays. Insertions at a specific index are 
  not supported.
* Be careful, appending or prepending an array variable to an array leads to a nested array
  (for more details please see this [issue](https://github.com/kubernetes-sigs/cluster-api/issues/5944)).

**Setting variable values in the Cluster**

> The user can now provide a value for the variable in the Cluster.

Reviewer 2: ```suggestion
After creating a ClusterClass with variable definition, the user will be allowed to provide a value for the variable in the Cluster as in the example below.
```

Author: Thx! I used your variant, but changed `will be allowed to provide` to `can now provide`

```yaml
apiVersion: cluster.x-k8s.io/v1beta1
kind: ClusterClass
metadata:
  name: docker-clusterclass-v0.1.0
spec:
  ...
  patches:
  - name: customImage
    description: "Sets the container image that is used for running dockerMachines."
    definitions:
    - selector:
        apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
        kind: DockerMachineTemplate
        matchResources:
          machineDeploymentClass:
            names:
            - default-worker
      jsonPatches:
      - op: add
        path: /spec/template/spec/customImage
        valueFrom:
          template: |
            kindest/node:{{ .builtin.machineDeployment.version }}
```

### JSON patch values

We already saw above that it's possible to use variable values in JSON patches. It's also 
possible to calculate values via Go templating or to use hard-coded values.

```yaml
apiVersion: cluster.x-k8s.io/v1beta1
kind: ClusterClass
metadata:
  name: docker-clusterclass-v0.1.0
spec:
  ...
  patches:
  - name: etcdImageTag
    definitions:
    - selector:
      ...
      jsonPatches:
      - op: add
        path: /spec/template/spec/kubeadmConfigSpec/clusterConfiguration/etcd
        valueFrom:

>           # This template is first rendered with Go templating, then parsed by 

Reviewer: It might be a good idea to explain step by step what the operation is here, and what the outcome of the template/patch will be.

Author: I added an example. I don't want to dive to deep into what is going on in the controller.

spec:
  ...
      jsonPatches:
      - op: add
        path: /spec/template/spec/httpProxy/url
        valueFrom:
          # Use the url field of the httpProxy variable.
          variable: httpProxy.url
      - op: add
        path: /spec/template/spec/dnsServers
        valueFrom:
          # Use the entire dnsServers array.
          variable: dnsServers
      - op: add
        path: /spec/template/spec/dnsServer
        valueFrom:
          # Use the first item of the dnsServers array.
          variable: dnsServers[0]
```

### Optional patches

Patches can also be conditionally enabled. This can be done by configuring a Go template via `enabledIf`. 
The patch is then only applied if the Go template evaluates to `true`. In the following example the `httpProxy` 
patch is only applied if the `httpProxy` variable is set (and not empty).

```yaml
apiVersion: cluster.x-k8s.io/v1beta1
kind: ClusterClass
metadata:
  name: docker-clusterclass-v0.1.0
spec:
  ...
  variables:
  - name: httpProxy
    schema:
      openAPIV3Schema:
        type: string
  patches:
  - name: httpProxy
    enabledIf: "{{ if .httpProxy }}true{{end}}"
    definitions:
    ...  
```

**Tips & Tricks**:

> Hard-coded values can be used to test the impact of a patch during development, gradually roll out patches, etc. .

Reviewer: This section is really great!

Author: Thank you! :)

```yaml
    enabledIf: "{{ if eq "v1.21.1" .builtin.controlPlane.version }}true{{end}}"
```

### MachineDeployment variable overrides

If you want to use many variations of MachineDeployments in Clusters, you can either define 
a MachineDeployment class for every variation or you can define patches and variables to 
make a single MachineDeployment class more flexible.

In the following example we make the `instanceType` of a `AWSMachineTemplate` customizable.
First we define the `workerMachineType` variable and the corresponding patch:

```yaml
apiVersion: cluster.x-k8s.io/v1beta1
kind: ClusterClass
metadata:
  name: aws-clusterclass-v0.1.0
spec:
  ...
  variables:
  - name: workerMachineType
    required: true
    schema:
      openAPIV3Schema:
        type: string
        default: t3.large
  patches:
  - name: workerMachineType
    definitions:
    - selector:
        apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
        kind: AWSMachineTemplate
        matchResources:
          machineDeploymentClass:
            names:
            - default-worker
      jsonPatches:
      - op: replace

>         path: /spec/template/spec/instanceType

Reviewer: I think including the templates in these yamls would be great to show exactly where the patch is happening.

Author: Good point here, as it helps to illustrate the example and it doesn't add to much YAML. Done.

referenced templates of a ClusterClass. A concrete example would be to deploy Clusters in 
different AWS regions. In this case, every cluster needs a Cluster-specific value for 
`.spec.region` in `AWSCluster`. Use cases like this can be implemented with ClusterClass 
patches.

**Defining variables in the ClusterClass**

The following example shows how variables can be defined in the ClusterClass.
A variable definition specifies the name and the schema of a variable and if it is 
required. The schema defines how a variable is defaulted and validated. It supports 
a subset of the schema of CRDs. For more information please see the [godoc](https://doc.crds.dev/github.com/kubernetes-sigs/cluster-api/cluster.x-k8s.io/ClusterClass/v1beta1#spec-variables-schema-openAPIV3Schema).

```yaml
apiVersion: cluster.x-k8s.io/v1beta1
kind: ClusterClass
metadata:
  name: docker-clusterclass-v0.1.0
spec:
  ...
  variables:
  - name: imageRepository
    required: true
    schema:
      openAPIV3Schema:
        type: string
        description: ImageRepository is the container registry to pull images from.
        default: k8s.gcr.io
        example: k8s.gcr.io
```

> **Note**: The following basic types are supported: `string`, `integer`, `number` and `boolean`. Complex types 

Reviewer 2: What about using aside for notes (this is almost the standard in the book)

Author: Done.

            kind: KubeadmConfigTemplate
            name: docker-clusterclass-v0.1.0-default-worker
            namespace: default
        infrastructure:
          ref:
            apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
            kind: DockerMachineTemplate
            name: docker-clusterclass-v0.1.0-default-worker
            namespace: default
```

The following example shows a Cluster using this ClusterClass. In this case a `KubeadmControlPlane` 
with the corresponding `DockerMachineTemplate`, a `DockerCluster` and a `MachineDeployment` with 
the corresponding `KubeadmConfigTemplate` and `DockerMachineTemplate` will be created. This basic 
ClusterClass is already very flexible. Via the topology on the Cluster the following can be configured:
* `.spec.topology.version`: the Kubernetes version of the Cluster
* `.spec.topology.controlPlane`: ControlPlane replicas and their metadata
* `.spec.topology.workers`: MachineDeployments and their replicas, metadata and failure domain

<aside class="note">

It's possible to create MachineDeployments with different configurations, by
defining multiple MachineDeployment classes in the ClusterClass.

> </aside>

Reviewer 2: I would drop this note here given that it refers to ClusterClass while instead here we are talking about the Cluster

Author: Fine for me

referenced templates of a ClusterClass. A concrete example would be to deploy Clusters in 
different AWS regions. In this case, every cluster needs a Cluster-specific value for 
`.spec.region` in `AWSCluster`. Use cases like this can be implemented with ClusterClass 
patches.

**Defining variables in the ClusterClass**

The following example shows how variables can be defined in the ClusterClass.
A variable definition specifies the name and the schema of a variable and if it is 
required. The schema defines how a variable is defaulted and validated. It supports 
a subset of the schema of CRDs. For more information please see the [godoc](https://doc.crds.dev/github.com/kubernetes-sigs/cluster-api/cluster.x-k8s.io/ClusterClass/v1beta1#spec-variables-schema-openAPIV3Schema).

```yaml
apiVersion: cluster.x-k8s.io/v1beta1
kind: ClusterClass
metadata:
  name: docker-clusterclass-v0.1.0
spec:
  ...
  variables:
  - name: imageRepository
    required: true
    schema:
      openAPIV3Schema:
        type: string
        description: ImageRepository is the container registry to pull images from.
        default: k8s.gcr.io
        example: k8s.gcr.io
```

> <aside class="note">

Reviewer 2: ```suggestion
<aside class="note">
<h1>Supported types</h1>
```
I find that having a title in the note helps readability

Author: Good point!

</aside>

**Defining patches in the ClusterClass**

The variable can then be used in a patch to set a field on a template referenced in the ClusterClass.
The `selector` specifies on which template the patch should be applied. `jsonPatches` specifies which JSON 
patches should be applied to that template. In this case we set the `imageRepository` field of the 
`KubeadmControlPlaneTemplate` to the value of the variable `imageRepository`. For more information 
please see the [godoc](https://doc.crds.dev/github.com/kubernetes-sigs/cluster-api/cluster.x-k8s.io/ClusterClass/v1beta1#spec-patches-definitions).

```yaml
apiVersion: cluster.x-k8s.io/v1beta1
kind: ClusterClass
metadata:
  name: docker-clusterclass-v0.1.0
spec:
  ...
  patches:
  - name: imageRepository
    definitions:
    - selector:
        apiVersion: controlplane.cluster.x-k8s.io/v1beta1
        kind: KubeadmControlPlaneTemplate
        matchResources:
          controlPlane: true
      jsonPatches:
      - op: add
        path: /spec/template/spec/kubeadmConfigSpec/clusterConfiguration/imageRepository
        valueFrom:
          variable: imageRepository
```

> <aside class="note">

Reviewer 2: ```suggestion
<aside class="note">
<h1>Writing json patches</h1>
```

* Only fields below `/spec` can be patched.
* Only `add`, `remove` and `replace` operations are supported.
* It's only possible to append and prepend to arrays. Insertions at a specific index are 
  not supported.
* Be careful, appending or prepending an array variable to an array leads to a nested array
  (for more details please see this [issue](https://github.com/kubernetes-sigs/cluster-api/issues/5944)).

</aside>

**Setting variable values in the Cluster**

After creating a ClusterClass with a variable definition, the user can now provide a value for 
the variable in the Cluster as in the example below.

```yaml
apiVersion: cluster.x-k8s.io/v1beta1
kind: Cluster
metadata:
  name: my-docker-cluster
spec:
  topology:
    ...
    variables:
    - name: imageRepository
      value: k8s.gcr.io
```

<aside class="note">

If the user does not set the value, the variable is automatically added with its 
default value by the Cluster webhook, as specified in the corresponding variable 
definition in the ClusterClass.

</aside>

> ## Advanced features of ClusterClass with patches

Reviewer 2: Proposing a little order reshuffle of the use cases

- MachineDeployment variable overrides
- Builtin variables
- Complex variable types
- Using variables values in JSON patches
- Optional patches

Author: Fine for me

    definitions:
    - selector:
      ...
      jsonPatches:
      - op: add
        path: /spec/template/spec/kubeadmConfigSpec/clusterConfiguration/controllerManager/extraArgs/cluster-name
        valueFrom:
          variable: builtin.cluster.name
```

**Tips & Tricks**

Builtin variables can be used to dynamically calculate image names. The version used in the patch 
will always be the same as the one we set in the corresponding MachineDeployment (works the same way 
with `.builtin.controlPlane.version`).

```yaml
apiVersion: cluster.x-k8s.io/v1beta1
kind: ClusterClass
metadata:
  name: docker-clusterclass-v0.1.0
spec:
  ...
  patches:
  - name: customImage
    description: "Sets the container image that is used for running dockerMachines."
    definitions:
    - selector:
        apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
        kind: DockerMachineTemplate
        matchResources:
          machineDeploymentClass:
            names:
            - default-worker
      jsonPatches:
      - op: add
        path: /spec/template/spec/customImage
        valueFrom:
          template: |
            kindest/node:{{ .builtin.machineDeployment.version }}
```

> ### JSON patch values

Reviewer 2: ```suggestion
### Using variables values in JSON patches
```
